daemon/volumes.go

```diff
@@ -65,9 +65,21 @@ func (m mounts) parts(i int) int {
 // 2. Select the volumes mounted from another containers. Overrides previously configured mount point destination.
 // 3. Select the bind mounts set by the client. Overrides previously configured mount point destinations.
 // 4. Cleanup old volumes that are about to be reassigned.
-func (daemon *Daemon) registerMountPoints(container *container.Container, hostConfig *containertypes.HostConfig) error {
+func (daemon *Daemon) registerMountPoints(container *container.Container, hostConfig *containertypes.HostConfig) (retErr error) {
 	binds := map[string]bool{}
 	mountPoints := map[string]*volume.MountPoint{}
+	defer func() {
+		// clean up the container mountpoints once return with error
+		if retErr != nil && len(mountPoints) != 0 {
```

Comment:
@cpuguy83  if `registerMountPoints` returns error, the mountpoints which has been registered in `registerMountPoints` would not in `container.MountPoint`, the caller unwind this on error in https://github.com/docker/docker/blob/master/daemon/create.go#L107 based on `container.MountPoint`